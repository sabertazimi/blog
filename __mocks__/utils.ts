let navigatorClipboardMocked = false
const writeTextMock = jest.fn()

const mockNavigatorClipboard = () => {
  if (!navigatorClipboardMocked) {
    Object.defineProperty(global.navigator, 'clipboard', {
      value: {
        writeText: writeTextMock,
      },
    })

    navigatorClipboardMocked = true
  }

  return { writeTextMock }
}

export { mockNavigatorClipboard }
