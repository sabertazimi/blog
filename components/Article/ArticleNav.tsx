import FlexContainer from '@components/FlexContainer'
import { ArrowLeft, ArrowRight, Home } from '@components/Icons'
import Link from '@components/Link'
import { Bounce } from '@components/Motion'
import type { PostMeta } from '@types'

interface Props {
  prevPost: PostMeta['prevPost']
  nextPost: PostMeta['nextPost']
}

const ArticleNav = ({ prevPost, nextPost }: Props): JSX.Element => (
  <FlexContainer role="navigation" aria-label="footer-navigation">
    <Bounce className="mr-6 flex-1 md:mr-24">
      <Link
        href={prevPost ? `/post/${prevPost.slug}` : '/posts'}
        className="flex-container bg-gradient-primary w-full rounded-full py-8 text-4xl text-light"
      >
        {prevPost ? (
          <ArrowLeft aria-label="Prev" />
        ) : (
          <Home aria-label="Home" />
        )}
      </Link>
    </Bounce>
    <Bounce className="ml-6 flex-1 md:ml-24">
      <Link
        href={nextPost ? `/post/${nextPost.slug}` : '/posts'}
        className="flex-container bg-gradient-primary w-full rounded-full py-8 text-4xl text-light"
      >
        {nextPost ? (
          <ArrowRight aria-label="Next" />
        ) : (
          <Home aria-label="Home" />
        )}
      </Link>
    </Bounce>
  </FlexContainer>
)

export default ArticleNav
