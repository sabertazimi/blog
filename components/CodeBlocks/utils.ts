import parseNumericRange from 'parse-numeric-range'
import type { Language } from 'prism-react-renderer'

const normalizeCode = (code: string = '') => code.replace(/\n+$/, '')

function isLanguage(language: string): language is Language {
  return language.length > 0
}

function normalizeLanguage(languageClass?: string): Language {
  if (languageClass != null) {
    const language = languageClass.replace('language-', '')

    if (isLanguage(language))
      return language
  }

  return 'typescript'
}

function normalizeLanguageName(language: string): string {
  switch (language) {
    case 'html':
      return 'HTML'
    case 'xml':
      return 'XML'
    case 'yml':
    case 'yaml':
      return 'YAML'
    case 'css':
      return 'CSS'
    case 'json':
      return 'JSON'
    case 'md':
    case 'markdown':
      return 'Markdown'
    case 'js':
    case 'javascript':
      return 'JavaScript'
    case 'ts':
    case 'typescript':
      return 'TypeScript'
    case 'coffee':
    case 'coffeescript':
      return 'CoffeeScript'
    case 'jsx':
    case 'tsx':
      return 'React'
    case 'objc':
    case 'objectivec':
      return 'Objective-C'
    default:
      return language.charAt(0).toUpperCase() + language.slice(1)
  }
}

function normalizeLines(expression: string): Set<number> {
  return new Set(parseNumericRange(expression))
}

export {
  normalizeCode,
  normalizeLanguage,
  normalizeLanguageName,
  normalizeLines,
}
