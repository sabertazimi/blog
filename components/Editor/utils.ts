function languageToFilepath(language?: string): string {
  switch (language) {
    case 'css':
      return '/styles.css'
    case 'js':
    case 'javascript':
      return '/App.js'
    case 'ts':
    case 'typescript':
      return '/App.ts'
    case 'jsx':
      return '/App.jsx'
    case 'tsx':
      return '/App.tsx'
    case 'vue':
      return '/src/App.vue'
    case undefined:
    default:
      return '/App.tsx'
  }
}

function normalizeFilepath(filename?: string, language?: string): string {
  if (filename == null || filename === '')
    return languageToFilepath(language)

  if (!filename.startsWith('/'))
    return `/${filename}`

  return filename
}

export { normalizeFilepath }
