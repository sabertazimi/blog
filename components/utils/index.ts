import classNames from 'classnames';

export { classNames };
