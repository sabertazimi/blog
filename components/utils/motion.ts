export { AnimatePresence, motion } from 'framer-motion';
export type { MotionProps } from 'framer-motion';
