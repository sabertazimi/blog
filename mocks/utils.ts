let navigatorClipboardMocked = false
const writeTextMock = jest.fn()

function mockNavigatorClipboard() {
  if (!navigatorClipboardMocked) {
    Object.defineProperty(globalThis.navigator, 'clipboard', {
      value: {
        writeText: writeTextMock,
      },
    })

    navigatorClipboardMocked = true
  }

  return { writeTextMock }
}

export { mockNavigatorClipboard }
