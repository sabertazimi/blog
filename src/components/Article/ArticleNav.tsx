import { FlexContainer } from '@components';
import type { PostMetaType } from '@types';
import { Button } from 'antd';
import {
  ArrowLeftOutlined,
  ArrowRightOutlined,
  HomeOutlined,
} from '@ant-design/icons';
import classNames from 'classnames';
import { Link } from 'gatsby';
import React from 'react';

interface Props {
  post: PostMetaType;
}

const ArticleNav = ({ post: { prevPost, nextPost } }: Props): JSX.Element => (
  <FlexContainer className="justify-between" role="navigation">
    <Button
      className={classNames(
        'flex-container',
        'w-full h-24 mb-6',
        'text-2xl align-bottom rounded-none',
        'md:mb-0 md:w-5/12',
        'button-primary'
      )}
      size="large"
    >
      {prevPost ? (
        <ArrowLeftOutlined aria-label="Prev" />
      ) : (
        <HomeOutlined aria-label="Home" />
      )}
      <Link
        className="m-auto font-extrabold text-light"
        to={prevPost ? prevPost.slug : '/posts'}
      >
        {prevPost ? prevPost.title : 'Back to Home'}
      </Link>
    </Button>
    <Button
      className={classNames(
        'flex-container',
        'w-full h-24 mb-6',
        'text-2xl align-bottom rounded-none',
        'md:mb-0 md:w-5/12',
        'button-primary'
      )}
      size="large"
    >
      <Link
        className="m-auto font-extrabold text-light"
        to={nextPost ? nextPost.slug : '/posts'}
      >
        {nextPost ? nextPost.title : 'Back to Home'}
      </Link>
      {nextPost ? (
        <ArrowRightOutlined aria-label="Next" />
      ) : (
        <HomeOutlined aria-label="Home" />
      )}
    </Button>
  </FlexContainer>
);

export default ArticleNav;
