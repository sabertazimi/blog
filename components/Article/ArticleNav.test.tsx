import mockData from '@/mocks/data'
import { render } from '@/utils'
import ArticleNav from './ArticleNav'

describe('ArticleNav', () => {
  const mockBasePost = mockData.basePosts[0]
  const mockPost = mockData.posts[0]

  it('should render correctly (snapshot)', () => {
    const { container } = render(<ArticleNav prevPost={mockPost.prevPost} nextPost={mockPost.nextPost} />)

    expect(container).toMatchSnapshot()
  })

  it('should render correctly with partial data (snapshot)', () => {
    const { container } = render(<ArticleNav prevPost={mockBasePost.prevPost} nextPost={mockBasePost.nextPost} />)

    expect(container).toMatchSnapshot()
  })
})
