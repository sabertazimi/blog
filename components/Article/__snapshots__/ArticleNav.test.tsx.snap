// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleNav should render correctly (snapshot) 1`] = `
<div>
  <div
    aria-label="footer-navigation"
    class="flex-container p-auto container relative mx-auto my-0 h-full"
    role="navigation"
  >
    <div
      class="inline-flex mr-6 flex-1 md:mr-24"
      style="opacity: 0; transform: translateX(20px);"
      tabindex="0"
    >
      <a
        class="flex-container bg-gradient-primary w-full rounded-full py-8 text-4xl text-light"
        href="/post/2BasicNotes"
      >
        <span
          aria-label="Prev"
          class="anticon anticon-arrow-left"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="arrow-left"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M872 474H286.9l350.2-304c5.6-4.9 2.2-14-5.2-14h-88.5c-3.9 0-7.6 1.4-10.5 3.9L155 487.8a31.96 31.96 0 000 48.3L535.1 866c1.5 1.3 3.3 2 5.2 2h91.5c7.4 0 10.8-9.2 5.2-14L286.9 550H872c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
            />
          </svg>
        </span>
      </a>
    </div>
    <div
      class="inline-flex ml-6 flex-1 md:ml-24"
      style="opacity: 0; transform: translateX(20px);"
      tabindex="0"
    >
      <a
        class="flex-container bg-gradient-primary w-full rounded-full py-8 text-4xl text-light"
        href="/post/0BasicNotes"
      >
        <span
          aria-label="Next"
          class="anticon anticon-arrow-right"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="arrow-right"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M869 487.8L491.2 159.9c-2.9-2.5-6.6-3.9-10.5-3.9h-88.5c-7.4 0-10.8 9.2-5.2 14l350.2 304H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h585.1L386.9 854c-5.6 4.9-2.2 14 5.2 14h91.5c1.9 0 3.8-.7 5.2-2L869 536.2a32.07 32.07 0 000-48.4z"
            />
          </svg>
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`ArticleNav should render correctly with partial data (snapshot) 1`] = `
<div>
  <div
    aria-label="footer-navigation"
    class="flex-container p-auto container relative mx-auto my-0 h-full"
    role="navigation"
  >
    <div
      class="inline-flex mr-6 flex-1 md:mr-24"
      style="opacity: 0; transform: translateX(20px);"
      tabindex="0"
    >
      <a
        class="flex-container bg-gradient-primary w-full rounded-full py-8 text-4xl text-light"
        href="/posts"
      >
        <span
          aria-label="Home"
          class="anticon anticon-home"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="home"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M946.5 505L560.1 118.8l-25.9-25.9a31.5 31.5 0 00-44.4 0L77.5 505a63.9 63.9 0 00-18.8 46c.4 35.2 29.7 63.3 64.9 63.3h42.5V940h691.8V614.3h43.4c17.1 0 33.2-6.7 45.3-18.8a63.6 63.6 0 0018.7-45.3c0-17-6.7-33.1-18.8-45.2zM568 868H456V664h112v204zm217.9-325.7V868H632V640c0-22.1-17.9-40-40-40H432c-22.1 0-40 17.9-40 40v228H238.1V542.3h-96l370-369.7 23.1 23.1L882 542.3h-96.1z"
            />
          </svg>
        </span>
      </a>
    </div>
    <div
      class="inline-flex ml-6 flex-1 md:ml-24"
      style="opacity: 0; transform: translateX(20px);"
      tabindex="0"
    >
      <a
        class="flex-container bg-gradient-primary w-full rounded-full py-8 text-4xl text-light"
        href="/posts"
      >
        <span
          aria-label="Home"
          class="anticon anticon-home"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="home"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M946.5 505L560.1 118.8l-25.9-25.9a31.5 31.5 0 00-44.4 0L77.5 505a63.9 63.9 0 00-18.8 46c.4 35.2 29.7 63.3 64.9 63.3h42.5V940h691.8V614.3h43.4c17.1 0 33.2-6.7 45.3-18.8a63.6 63.6 0 0018.7-45.3c0-17-6.7-33.1-18.8-45.2zM568 868H456V664h112v204zm217.9-325.7V868H632V640c0-22.1-17.9-40-40-40H432c-22.1 0-40 17.9-40 40v228H238.1V542.3h-96l370-369.7 23.1 23.1L882 542.3h-96.1z"
            />
          </svg>
        </span>
      </a>
    </div>
  </div>
</div>
`;
